File: DIDP/src/main/java/com/lizbyu/structural/proxy/Proxy.java
package com.lizbyu.structural.proxy;

import com.lizbyu.utils.log.LogUtils;

public class Proxy implements Subject {
    private static final String TAG = "Proxy";
    private final Subject mRealSubject;

    public Proxy() {
        this.mRealSubject = new RealSubject();
    }

    @Override
    public void request() {
        preRequest();
        mRealSubject.request();
        postRequest();
    }

    private void preRequest() {
        LogUtils.d(TAG, "pre request ...");
    }

    private void postRequest() {
        LogUtils.d(TAG, "post request ...");
    }
}

File: Section5/src/structural/proxy/Proxy.kt
package structural.proxy

class Proxy : Subject {
    private var realSubject : Subject? = null

    constructor(realSubject: Subject) {
        this.realSubject = realSubject
    }

    private fun doFirst() {
        println("Proxy do first ...")
    }

    override fun request() {
        doFirst()
        realSubject!!.request()
        doLast()
    }

    private fun doLast() {
        println("Proxy do last ...")
    }

}